third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -2813,112 +2825,235 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
+                        "field `{}` not declared in struct `{}`",
+                        field_name.display(self.symbol_pool()),
                         struct_name.display(self.parent.parent.env)
                     ),
                 );
-                None
+                succeed = false;
             }
-        } else {
+        }
+        if !fields_not_covered.is_empty() {
             self.error(
-                &struct_name_loc,
+                loc,
                 &format!(
-                    "undeclared struct `{}`",
-                    struct_name.display(self.parent.parent.env)
+                    "missing fields {}",
+                    fields_not_covered
+                        .iter()
+                        .map(|n| format!("`{}`", n.display(self.symbol_pool())))
+                        .join(", ")
                 ),
             );
+            succeed = false;
+        }
+        if succeed {
+            Some(())
+        } else {
             None
         }
     }
 
+    // given pack<S>{ f_p(1): e_1, ... }, where p is a permutation of the fields
+    // return the struct id of S,
+    // (.f_p(1), e_1'), (.f_p(1), e_2') ... where e_i' is e_i translated
+    // and
+    // .f_q(1), .f_q(2), ... where q is the inverse of p
+    fn translate_exp_fields(
+        &mut self,
+        loc: &Loc,
+        maccess: &EA::ModuleAccess,
+        generics: &Option<Vec<EA::Type>>,
+        fields: &EA::Fields<EA::Exp>,
+    ) -> Option<(
+        QualifiedInstId<StructId>,
+        Vec<(Pattern, ExpData)>,
+        Vec<ExpData>,
+    )> {
+        let struct_name = self.parent.module_access_to_qualified(maccess);
+        let struct_name_loc = self.to_loc(&maccess.loc);
+        let generics = generics.as_ref().map(|ts| self.translate_types(ts));
+        let struct_entry = self.get_struct_report_undeclared(&struct_name, &struct_name_loc)?;
+        let instantiation =
+            self.make_instantiation_or_report(loc, struct_entry.type_params.len(), generics)?;
+        let field_decls = self
+            .get_field_decls_for_pack_unpack(&struct_entry, &struct_name, &struct_name_loc)?
+            .clone();
+
+        self.check_missing_or_undeclared_fields(loc, struct_name, &field_decls, fields)?;
+
+        // maps i to (x_i, e_i) where i is the expression idx, i.e., idx of the field exp in the pack exp
+        let mut bindings = BTreeMap::new();
+        // maps i to x_i where i is the field definition idx, i.e., idx of the field in the struct definition
+        let mut args = BTreeMap::new();
+        for (_, name, (exp_idx, field_exp)) in fields.iter() {
+            let field_name = self.symbol_pool().make(name);
+            let (_, def_idx, field_ty) = field_decls.get(&field_name).unwrap();
+            let field_ty = field_ty.instantiate(&instantiation);
+            let translated_field_exp = self.translate_exp(field_exp, &field_ty);
+            // starts with . to avoid shadowing
```

Comment:
We need to document somewhere if we're using special symbol names to mean something.  I'm having the same problem with inlining (avoiding shadowing existing variables) and we should avoid potential conflicts with future name-mangling or something.